### ISSUE TITLE
Comment for bounded_lru_cache_wrapper seems incorrect

### ISSUE THREAD

**Issue by YouJiacheng** [python/cpython#94879]
https://github.com/python/cpython/blob/c22f134211743cd5ad14cec1dd4f527bee542b4c/Modules/_functoolsmodule.c#L975-L976
However the `__eq__` method call actually could be made for the deletion of the oldest entry.
```python
from functools import lru_cache

class A:
    def __init__(self, name):
        self.name = name
        self.counter = []

    def __hash__(self):
        return 0 # mock hash collision
    
    def __eq__(self, __o: 'A'):
        self.counter.append(__o.name)
        return self is __o

@lru_cache(maxsize=2)
def foo(_):
    return 0

a, b, c = A('a'), A('b'), A('c')

foo(a)
foo(b)

# make 'b' be the oldest-used entry
# but not the oldest-inserted entry
foo(a)
a.counter.clear()

foo(c) # will evict 'b'

# The assertion shows that
# 'a' was compared with 'b' for the deletion of 'b'
assert 'b' in a.counter
print(a.counter) # ['c', 'c', 'b', 'c']
```
It seems that reentrant here might cause the assertion in L1092 failed:
https://github.com/python/cpython/blob/c22f134211743cd5ad14cec1dd4f527bee542b4c/Modules/_functoolsmodule.c#L1092-L1099
Since link has been extracted but hasn't been popped from cache dict, link list could be empty while cache dict is full.

**Comment by rhettinger** [comment]
Serhiy, do you want to look at this?  The deletion step is almost always an identity match, but it is possible to trigger an ``__eq__()`` call if two unequal objects have the same hash, if the dict.pop() step tests the other, non-identical object first, and if that object has a pure python  ``__eq__()`` method.

Offhand, the only only non-invasive fix I can think of is to replace the call to ``_PyDict_Pop_KnownHash()`` with a new ``_PyDict_Pop_IdenticalWithKnownHash()`` that never calls ``__eq__()`` and only deletes the exact key we're looking for.

**Comment by serhiy-storchaka** [comment]
The assertion in L1092 cannot fail, because we go here only if the following condition is false:
https://github.com/python/cpython/blob/c22f134211743cd5ad14cec1dd4f527bee542b4c/Modules/_functoolsmodule.c#L1048-L1049

In other words, when both `PyDict_GET_SIZE(self->cache) >= self->maxsize` and `self->root.next != &self->root` are true.

I agree with Raymond, that using a new `_PyDict_Pop_IdenticalWithKnownHash()` is the simplest way to guarantee that `__eq__()` is not called here. We can also make `_PyDict_Pop_KnownHash()` testing all keys with the same hash using only identity in the first pass. It will save one C API functions, but will make the code more complex internally.

But what problem it will actually solve besides not exactly correct comment? Is it worth?

**Comment by rhettinger** [comment]
> Is it worth?

The OP found a case where the code could allow a reentrant call in the middle of pop operation.  The question is whether that cache is in a consistent state at that point so that the other call works correctly and don't mangle the cache invariants.   If so, then this is worth fixing.  If not, then the comment can be updated with an explanation for why we don't think it is a problem.

**Comment by rhettinger** [comment]
Tim, do you have an interest in looking at this with us.  If not, unassign.

**Comment by YouJiacheng** [comment]
@serhiy-storchaka Considering this condition, the thing seems worse: 
Let's say there are enough reentrants, then this condition will be true every 2 reentrants (after link list is empty while cache dict is full), and the net effect of every 2 reentrants is the cache dict size will increase 1, which is not recoverable after reentrant stop, since the "false" reentrant will pop the item inserted by "true" reentrant, and insert its own.
Is my understanding correct?

**Comment by serhiy-storchaka** [comment]
Yes, you are right, although such scenario should be very rare on practice. First you need `maxsize` reentrants to empty the linked list, and only after this every second reentrant will increase the cache size. For example, if `maxsize`=200, 300 reentrant calls (all with unique arguments) will increase the cache size to 250. But a cache with `maxsize`=200 is just useless for the case where you have 300 simultaneous calls with different arguments.

The limit on the cache size can be made more strict if continue to delete the oldest key while the size exceeds the limit.

**Comment by serhiy-storchaka** [comment]
I have found other related issues:

* In Python implementation you can get KeyError from `del cache[oldkey]`.
* In C implementation, short time after removing the oldest item from the linked list and before removing it from the dict, the dict contains the link object with inconsistent next/prev links. The reentrant call can find the half-removed item in the dict and try to move it to the top of the linked list. If it was only one just removed from the linked list item, nothing bad happens (although we can add code to detect such case), but if several items are in such state (removed from the linked list, having inconsistent next/prev links, still in the dict), it can make mess with the linked list.

These issues are worth to fix.

But the fact that the size of the dict or the linked list can exceed `maxsize` in very rare circumstances is not a big deal if it is only for short time and the following random operations will return the situation to normal. It only possible if we have more than `maxsize` simultaneous or reentrant operations, and it already consumes O(maxsize) memory at once on stacks and in frames.

**Comment by serhiy-storchaka** [comment]
Example of KeyError for maxsize=2:

* Add keys K1 and K2 to fill the cache.
* Add key K3, but after updating the linked list (it is now `[K2, K3]`) and before adding K3 to the cache dict re-enter and:
  * Add key K4, but before updating the cache re-enter and:
    * Add key K2, which is found in the cache. It reorders the linked list as `[K3, K2]`.
  * Continue with K4. Remove the oldest key (K3) from the cache dict. But it was not added yet.
